tests-integration/poller_test.go

```diff
@@ -428,3 +428,32 @@ func TestPollersCanBeResumedAfterExpiry(t *testing.T) {
 	m.MatchResponse(t, res, m.MatchAccountData([]json.RawMessage{accdata}, nil))
 
 }
+
+// Regression test for https://github.com/matrix-org/sliding-sync/issues/287#issuecomment-1706522718
+func TestPollerExpiryEnsurePollingRace(t *testing.T) {
+	pqString := testutils.PrepareDBConnectionString()
+	v2 := runTestV2Server(t)
+	defer v2.close()
+	v3 := runTestServer(t, v2, pqString)
+	defer v3.close()
+
+	v2.addAccount(t, alice, aliceToken)
+
+	// Arrange the following:
+	// 1. A request arrives from an unknown token.
+	// 2. The API makes a /whoami lookup for the new token. That returns without error.
+	// 3. The old token expires.
+	// 4. The poller tries to call /sync but finds that the token has expired.
```

Comment:
This never fires off V2InitialSyncComplete. (Indeed it seems to get stuck in an infinite loop, because [401 is considered a temporary error](https://github.com/matrix-org/sliding-sync/blob/ff7120245a9f53f751f2c5642115755ca0237efb/sync2/poller.go#L572-L583). Perhaps we ought to abort if e.g. failCount > 5.)

Taking a step back, the behaviour we want to replicate is that EnsurePolling returns without there being an active poller. For this to happen we either hit [here](https://github.com/matrix-org/sliding-sync/blob/ff7120245a9f53f751f2c5642115755ca0237efb/sync2/poller.go#L255-L265) or the else branch [here](https://github.com/matrix-org/sliding-sync/blob/ff7120245a9f53f751f2c5642115755ca0237efb/sync2/poller.go#L293-L297). Maybe we can chase log lines to see which of these happens.

```
18:43:04 INF Received connection from unknown access token, querying with homeserver
18:43:04 TRC checking poller exists and is running conn= device=alice_TestPollerExpiryEnsurePollingRace_device pid={"DeviceID":"alice_TestPollerExpiryEnsurePollingRace_device","UserID":"@alice:localhost"} user=@alice:localhost
18:43:04 INF EnsurePolling: new request device_id=alice_TestPollerExpiryEnsurePollingRace_device user_id=@alice:localhost
18:43:04 INF EnsurePolling: preprocessing done device_id=alice_TestPollerExpiryEnsurePollingRace_device user_id=@alice:localhost
18:43:04 INF Poller: v2 poll loop started device_id=alice_TestPollerExpiryEnsurePollingRace_device since= user_id=@alice:localhost
    poller_test.go:454: Alice's token expires.
    poller_test.go:454: Alice's token expires.
18:43:04 WRN Poller: sync v2 poll returned temporary error error="DoSyncV2: request failed: Get \"http://127.0.0.1:44527/_matrix/client/r0/sync?timeout=0&filter=%7B%22room%22%3A%7B%22timeline%22%3A%7B%22limit%22%3A1%7D%7D%7D\": EOF" code=0 device_id=alice_TestPollerExpiryEnsurePollingRace_device user_id=@alice:localhost
18:43:04 WRN Poller: waiting before next poll device_id=alice_TestPollerExpiryEnsurePollingRace_device duration=3s fail-count=1 user_id=@alice:localhost
    poller_test.go:454: Alice's token expires.
18:43:07 WRN Poller: sync v2 poll returned temporary error error="DoSyncV2: request failed: Get \"http://127.0.0.1:44527/_matrix/client/r0/sync?timeout=0&filter=%7B%22room%22%3A%7B%22timeline%22%3A%7B%22limit%22%3A1%7D%7D%7D\": EOF" code=0 device_id=alice_TestPollerExpiryEnsurePollingRace_device user_id=@alice:localhost
18:43:07 WRN Poller: waiting before next poll device_id=alice_TestPollerExpiryEnsurePollingRace_device duration=3s fail-count=2 user_id=@alice:localhost
    poller_test.go:454: Alice's token expires.
18:43:10 WRN Poller: sync v2 poll returned temporary error error="DoSyncV2: request failed: Get \"http://127.0.0.1:44527/_matrix/client/r0/sync?timeout=0&filter=%7B%22room%22%3A%7B%22timeline%22%3A%7B%22limit%22%3A1%7D%7D%7D\": EOF" code=0 device_id=alice_TestPollerExpiryEnsurePollingRace_device user_id=@alice:localhost
18:43:10 WRN Poller: waiting before next poll device_id=alice_TestPollerExpiryEnsurePollingRace_device duration=3s fail-count=3 user_id=@alice:localhost

Process finished with the exit code 137 (interrupted by signal 9: SIGKILL)
```

Reply:
> For this to happen we either hit [here](https://github.com/matrix-org/sliding-sync/blob/ff7120245a9f53f751f2c5642115755ca0237efb/sync2/poller.go#L255-L265) or the else branch [here](https://github.com/matrix-org/sliding-sync/blob/ff7120245a9f53f751f2c5642115755ca0237efb/sync2/poller.go#L293-L297).

I think my money's on the former.

Reply:
Oh, I was invalidating the token improperly. I think it's better now.

Reply:
Also it wasn't returning a 401, it was `code=0` with an EOF.